### PR TITLE
feat[WIP]: Basic Auth

### PR DESCRIPTION
**Pull request by saadahmadhere**
This PR includes the following features:
1. User can log in with test credentials. (Please note that the use of local Storage & post request to handle Auth is not included in this PR).
2. If not logged in, the user is redirected to the login page when she wants to access any private route.
3. If not logged in, the user gets redirected to the login page when she wants to add an item to the Cart/Wishlist.
4. An Error message is displayed when not logging in using the test credentials.
5. The User is redirected back to the previous route which she wanted to go to, once she logs in successfully.
6. After logging out, she gets redirected back to the homepage.

### CODE REVIEW

File: src/Components/Card/Card.jsx [hunk 1]
```diff
@@ -1,16 +1,22 @@
...
   const addToCartHandler = (product) => {
-    dispatch({ type: "ADD_TO_CART", payload: product });
+    isLoggedIn
+      ? dispatch({ type: "ADD_TO_CART", payload: product })
+      : navigate("/login");
   };
 
   const addToWishListHandler = (product) => {
-    dispatch({ type: "ADD_TO_WISHLIST", payload: product });
+    isLoggedIn
```
Comment (lannisterrr) [inline src/Components/Card/Card.jsx]: you can also check for loggedin in the API call iteself.

File: src/Screens/Login Screen/Login.jsx [hunk 2]
```diff
@@ -1,19 +1,65 @@
...
+  };
+
+  const showPasswordHandler = (e) => {
+    e.preventDefault();
+    setShowPassword((prev) => !prev);
+  };
+
+  const submitHandler = (e) => {
+    e.preventDefault();
+    if (password === "adarshbalika" && email === "adarshbalika@gmail.com") {
```
Comment (lannisterrr) [inline src/Screens/Login Screen/Login.jsx]: I think the api call will check the test credentials internally , you don't have to do this explicitly.

**Review by lannisterrr — COMMENTED** [review]
Good code structure , some suggesstions made.

File: src/Components/Navbar/Navbar.jsx [hunk 3]
```diff
@@ -30,15 +38,26 @@ const Navbar = () => {
...
-              Login
-            </Link>
-          </li>
-          <li>
-            <Link to="/signUp" className="btn btn_outline btn_logout btn_sm">
-              Signup
-            </Link>
+            {isLoggedIn ? (
+              <button
+                to="/login"
```
Comment (shivampandey0) [inline src/Components/Navbar/Navbar.jsx]: `button` element doesn't have `to` attribute. Maybe you meant `Link` here?

**Review by shuklahimanshu81 — APPROVED** [review]
Good use of short circuiting . Code structure is clean . Good to merge.

**Review by shivampandey0 — CHANGES_REQUESTED** [review]
On `Login.jsx` fix the `submitHandler` to make the login request to the API with test credentials.

File: src/Screens/Login Screen/Login.jsx (hunk 2 above)
Comment (shivampandey0) [reply src/Screens/Login Screen/Login.jsx]: yeah, this needs to be fixed before merging.

File: src/Components/Card/Card.jsx (hunk 1 above)
Comment (saadahmadhere) [reply src/Components/Card/Card.jsx]: fixed in https://github.com/saadahmadhere/elmart/pull/13/commits/bd864a639437856ba88449d9f967a1624268009d

File: src/Components/Navbar/Navbar.jsx (hunk 3 above)
Comment (saadahmadhere) [reply src/Components/Navbar/Navbar.jsx]: Fixed in https://github.com/saadahmadhere/elmart/pull/13/commits/bd864a639437856ba88449d9f967a1624268009d

File: src/Screens/Login Screen/Login.jsx (hunk 2 above)
Comment (saadahmadhere) [reply src/Screens/Login Screen/Login.jsx]: fixed in https://github.com/saadahmadhere/elmart/pull/13/commits/e14d398edddba37855b47127a491c1dc54370132

File: src/Components/Card/Card.jsx [hunk 4]
```diff
@@ -1,26 +1,29 @@
...
+  const addToWishListHandler = () => {
+    userState.token
+      ? addToWishlist(product, userState.token, dispatchUserState)
       : navigate("/login");
   };
 
-  const removeFromWishlistHandler = (product) => {
-    dispatch({ type: "REMOVE_FROM_WISHLIST", payload: product });
+  const removeFromWishlistHandler = () => {
+    removeFromWishlist(product, userState.token, dispatchUserState);
```
Comment (shivampandey0) [inline src/Components/Card/Card.jsx]: `removeFromWishlist` method only needs `_id` from product, no need to pass the complete object here.
```suggestion
    removeFromWishlist(product._id, userState.token, dispatchUserState);
```

File: src/Components/CardHorizontal/CardHorizontal.jsx [hunk 5]
```diff
@@ -1,24 +1,31 @@
...
+      userState.token,
+      dispatchUserState,
+      e.target.value
+    );
   };
 
-  const removeCartItemHandler = (_id) => {
-    dispatch({ type: "REMOVE_FROM_CART", payload: _id });
+  const removeCartItemHandler = () => {
+    removeFromCart(product, userState.token, dispatchUserState);
```
Comment (shivampandey0) [inline src/Components/CardHorizontal/CardHorizontal.jsx]: method only needs `_id`
```suggestion
    removeFromCart(product._id, userState.token, dispatchUserState);
```

File: src/Screens/CartScreen/CartScreen.jsx [hunk 6]
```diff
@@ -1,14 +1,15 @@
...
 const CartScreen = () => {
   const {
-    state: { cart },
-  } = useData();
+    userState: {
+      userData: { cart },
+    },
+  } = useAuth();
 
   const [totalPrice, setTotalPrice] = useState(0);
```
Comment (shivampandey0) [inline src/Screens/CartScreen/CartScreen.jsx]: would like to talk about the use of these two states here. maybe you don't need them.

**Review by shivampandey0 — COMMENTED** [review]
Excellent abstraction of all the network request-related code in one place.
good use of semantic HTML tags.

- Added some suggestions

File: src/Screens/CartScreen/CartScreen.jsx (hunk 6 above)
Comment (saadahmadhere) [reply src/Screens/CartScreen/CartScreen.jsx]: fixed in https://github.com/saadahmadhere/elmart/pull/13/commits/d02b1a8531f501320f11b2d2f643a74250f96187

File: src/Components/CardHorizontal/CardHorizontal.jsx (hunk 5 above)
Comment (saadahmadhere) [reply src/Components/CardHorizontal/CardHorizontal.jsx]: fixed in https://github.com/saadahmadhere/elmart/pull/13/commits/1b5c9c8caec12d29427cb645b878d98bdb564977

File: src/Components/Card/Card.jsx (hunk 4 above)
Comment (saadahmadhere) [reply src/Components/Card/Card.jsx]: fixed in https://github.com/saadahmadhere/elmart/pull/13/commits/1b5c9c8caec12d29427cb645b878d98bdb564977

**Review by shivampandey0 — APPROVED** [review]

**Pull request merged**